cmse/cmse.md

```diff
@@ -0,0 +1,1725 @@
+are returned.
+
+When executed in the secure state the result of this instruction is extended to
+return the SAU and IDAU configurations at the specific address. 
+
+The MPU is banked between the two security states. The optional `A` flag makes
+the `TT` instruction read the MPU of the non-secure state when the `TT`
+instruction is executed from the secure state.
+
+The TT instruction is used to check the access permissions that different
+security states and privilege levels have on memory at a specified address.
+
+##	Secure code requirements
+
+To prevent secure code and data from being accessed from non-secure state,
+secure code must meet at least the requirements listed in this section. The
+responsibility for meeting these security requirements is shared between
+hardware, toolchain and software developer. The remainder of this document
+specifies requirements a toolchain must meet to enable C programmers to meet
+these security requirements.
+
+###	Information leakage
+
+Information leakage from the secure state to the non-secure state may occur
+through parts of the system that are not banked between the security states.
+The unbanked registers that are accessible by software are:
+
+*	General purpose registers except for the stack pointer (`r0-r12`, `r14-r15`).
+*	Floating point registers (`S0-S31`, `D0-D15`).
+*	The `N`, `Z`, `C`, `V`, `Q`, and `GE` bits of the `APSR` register.
+*	The `FPSCR` register.
+
+> 1 Secure code must clear secret information
+> from unbanked registers before initiating a transition from secure to
+> non-secure state.
+
+###	Non-secure memory access
+When secure code needs to access non-secure memory using an address calculated
+by the non-secure state, it cannot trust that the address lies in a non-secure
+memory region. Furthermore, the MPU is banked between the security states.
+Therefore secure and non-secure code might have different access rights to
+non-secure memory.
+
+> 2	Secure code that accesses non-secure memory on behalf of the non-secure
+> state must only do so if the non-secure state has permission to perform the
+> same access itself.
+
+The secure code can use the `TT` instruction to check non-secure memory
+permissions.
+
+> 3	Secure code must not access non-secure memory unless it does so on behalf
+> of the non-secure state.
+
+Data belonging to secure code must reside in secure memory.
+
+###	Volatility of non-secure memory
+
+Non-secure memory can be changed asynchronously to the execution of secure code.
+There are two causes:
+
+*	Interrupts handled in non-secure state can change non-secure memory.
+*	The debug interface can be used to change non-secure memory.
+
+There can be unexpected consequences when secure code accesses non-secure memory:
+
+```
+  int array[N]
+  void foo(int *p) {
+      if (*p >= 0 && *p < N) {
+          // non-secure memory (*p) is changed at this point
+          array[*p] = 0;
+      }
+  }
```

Comment:
There is no need to indent the code. Please start from column zero. I am amrking only this one but please apply it to all the code in the specs.
```suggestion
int array[N]
void foo(int *p) {
    if (*p >= 0 && *p < N) {
        // non-secure memory (*p) is changed at this point
        array[*p] = 0;
    }
}
```